File: samples/java/SupportChat/Server/Session/src/main/java/systems/zlink/samples/supportchat/server/session/sessions/SupportChatSession.java
package systems.zlink.samples.supportchat.server.session.sessions;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkStreamError;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.configuration.SampleTimings;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class SupportChatSession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkClient channels;
    private ZLinkSessionActor identityActor;
    private String identityActorId = "";
    private String identityDisplayName = "";
    private String identityRole = "";

    public SupportChatSession(ZLinkSessionContext context, ZLinkClient channels) {
        this.context = context;
        this.channels = channels;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        CompletableFuture<?>[] notifications =
                context.actors().bound().stream()
                        .map(actor -> actor.notifyDisconnected().toCompletableFuture())
                        .toArray(CompletableFuture[]::new);
        return CompletableFuture.allOf(notifications);
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [start:doc-sc-session-dispatch]
    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        return switch (dispatch.packetName()) {
            case "AuthenticateReq" -> authenticate(payload.decode(Messages.AuthenticateReq.class));
            case "JoinConversationReq" -> joinConversation(dispatch, payload);
            default -> relayConversationPacket(dispatch, payload);
        };
    }

    // --8<-- [end:doc-sc-session-dispatch]

    // --8<-- [start:doc-sc-session-auth]
    private CompletionStage<Void> authenticate(Messages.AuthenticateReq request) {
        return channels.requestToChannel(
                        SampleNames.ApiChannel,
                        new Messages.AuthenticateUserReq(request.accessToken()))
                .timeout(SampleTimings.RequestTimeout)
                .submit(Messages.AuthenticateUserRes.class)
                .thenCompose(
                        authenticated -> {
                            if (!authenticated.accepted()
                                    || authenticated.actorId() == null
                                    || authenticated.displayName() == null
                                    || authenticated.role() == null) {
                                throw new IllegalStateException(
                                        authenticated.reason() == null
                                                ? "SupportChat authentication failed"
                                                : authenticated.reason());
                            }
                            identityActorId = authenticated.actorId();
                            identityDisplayName = authenticated.displayName();
                            identityRole = authenticated.role();
                            return channels.requestToChannel(
                                            SampleNames.SupportChannel,
                                            new Messages.EnsureSupportUserActorReq(
                                                    identityActorId,
                                                    identityDisplayName,
                                                    identityRole,
                                                    identityActorId))
                                    .timeout(SampleTimings.RequestTimeout)
                                    .submit(Messages.EnsureSupportUserActorRes.class);
                        })
                .thenCompose(ensured -> bindOrGet(ensured.actor().toActorRef()))
                .thenAccept(
                        actor -> {
                            identityActor = actor;
                            context.client()
                                    .reply(
                                            new Messages.AuthenticateRes(
                                                    identityActorId,
                                                    identityDisplayName,
                                                    identityRole))
                                    .submit();
                        });
    }

    // --8<-- [end:doc-sc-session-auth]

    // --8<-- [start:doc-sc-agent-join]
    private CompletionStage<Void> joinConversation(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        Messages.JoinConversationReq request = payload.decode(Messages.JoinConversationReq.class);
        if (request.conversationId() == null || request.conversationId().isBlank()) {
            throw new IllegalStateException("Conversation Join is missing conversationId");
        }
        if (SampleNames.Roles.Customer.equals(identityRole)) {
            return requireIdentityActor().relay(dispatch, payload).thenApply(ignored -> null);
        }
        return channels.requestToChannel(
                        SampleNames.SupportChannel,
                        new Messages.EnsureAgentConversationReq(
                                identityActorId, identityDisplayName, request.conversationId()))
                .timeout(SampleTimings.RequestTimeout)
                .submit(Messages.EnsureAgentConversationRes.class)
                .thenCompose(ensured -> bindOrGet(ensured.actor().toActorRef()))
                .thenCompose(actor -> actor.relay(dispatch, payload));
    }

    // --8<-- [end:doc-sc-agent-join]

    private CompletionStage<Void> relayConversationPacket(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        // --8<-- [start:doc-sc-actor-relay]
        ZLinkSessionActor target = dispatch.actor();
        if (target == null) {
            target = requireIdentityActor();
        }
        // --8<-- [end:doc-sc-actor-relay]
        return target.relay(dispatch, payload).thenApply(ignored -> null);
    }

    private CompletionStage<ZLinkSessionActor> bindOrGet(ActorRef actorRef) {
        ZLinkSessionActor existing = context.actors().find(actorRef.actorId()).orElse(null);
        return existing == null
                ? context.actors().bind(actorRef)
                : CompletableFuture.completedFuture(existing);
    }

    private ZLinkSessionActor requireIdentityActor() {
        if (identityActor == null) {
            throw new IllegalStateException(
                    "Client must authenticate before sending conversation packets");
        }
        return identityActor;
    }
}

File: samples/kotlin/SupportChat/Server/Session/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/session/sessions/SupportChatSession.kt
package systems.zlink.samples.kotlin.supportchat.server.session.sessions

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.framework.kotlin.ZLinkSuspendingSession
import systems.zlink.framework.kotlin.bindOrGetActor
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSessionActor
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkStreamError
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AuthenticateReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AuthenticateRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AuthenticateUserReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AuthenticateUserRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureAgentConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureAgentConversationRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureSupportUserActorReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureSupportUserActorRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationRes

class SupportChatSession(
    private val context: ZLinkSessionContext,
    private val channels: ZLinkClient,
) : ZLinkSuspendingSession() {
    private val conversationActors = linkedMapOf<String, ZLinkSessionActor>()
    private var identityActor: ZLinkSessionActor? = null
    private var identityActorId: String = ""
    private var identityDisplayName: String = ""
    private var identityRole: String = ""

    override fun context(): ZLinkSessionContext = context

    override suspend fun onDisconnectedSuspending() {
        for (actor in context.actors().bound()) {
            actor.notifyDisconnected().await()
        }
    }

    override suspend fun onErrorSuspending(error: ZLinkStreamError) {}

    // --8<-- [start:doc-sc-session-dispatch]
    override suspend fun onDispatchSuspending(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        when (dispatch.packetName()) {
            "AuthenticateReq" -> authenticate(payload.decode(AuthenticateReq::class.java))
            "JoinConversationReq" -> joinConversation(dispatch, payload)
            else -> relayConversationPacket(dispatch, payload)
        }
    }

    // --8<-- [end:doc-sc-session-dispatch]

    private suspend fun authenticate(request: AuthenticateReq) {
        val authenticated =
            channels
                .requestToChannel(SampleNames.ApiChannel, AuthenticateUserReq(request.accessToken))
                .timeout(SampleTimings.RequestTimeout)
                .submit(AuthenticateUserRes::class.java)
                .await()
        if (
            !authenticated.accepted ||
                authenticated.actorId.isNullOrBlank() ||
                authenticated.displayName.isNullOrBlank() ||
                authenticated.role.isNullOrBlank()
        ) {
            throw IllegalStateException(
                authenticated.reason ?: "SupportChat authentication failed."
            )
        }
        val actorId =
            authenticated.actorId
                ?: throw IllegalStateException(
                    "SupportChat authentication did not return an actor id."
                )
        val displayName =
            authenticated.displayName
                ?: throw IllegalStateException(
                    "SupportChat authentication did not return a display name."
                )
        val role =
            authenticated.role
                ?: throw IllegalStateException("SupportChat authentication did not return a role.")

        // --8<-- [start:doc-sc-session-auth]
        val ensured =
            channels
                .requestToChannel(
                    SampleNames.SupportChannel,
                    EnsureSupportUserActorReq(
                        actorId = actorId,
                        displayName = displayName,
                        role = role,
                        participantId = actorId,
                    ),
                )
                .timeout(SampleTimings.RequestTimeout)
                .submit(EnsureSupportUserActorRes::class.java)
                .await()

        identityActor = context.actors().bindOrGetActor(ensured.actor.toActorRef())
        identityActorId = actorId
        identityDisplayName = displayName
        identityRole = role
        // --8<-- [end:doc-sc-session-auth]
        context.client().reply(AuthenticateRes(actorId, displayName, role)).submit()
    }

    private suspend fun joinConversation(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        if (identityRole == SupportChatRoles.Customer) {
            requireIdentityActor().relay(dispatch, payload).await()
            return
        }

        val conversationId = requireConversationId(dispatch)
        val existing = conversationActors[conversationId]
        if (existing != null) {
            existing.relay(dispatch, payload).await()
            return
        }

        // --8<-- [start:doc-sc-agent-join]
        val ensured =
            channels
                .requestToChannel(
                    SampleNames.SupportChannel,
                    EnsureAgentConversationReq(identityActorId, identityDisplayName, conversationId),
                )
                .timeout(SampleTimings.RequestTimeout)
                .submit(EnsureAgentConversationRes::class.java)
                .await()

        conversationActors[conversationId] =
            context.actors().bindOrGetActor(ensured.actor.toActorRef())
        logger.info(
            "session: agent joined conversation. roster={}, conversation={}",
            identityActorId,
            conversationId,
        )
        context.client().reply(JoinConversationRes(ensured.scheduled, ensured.state)).submit()
        // --8<-- [end:doc-sc-agent-join]
    }

    // --8<-- [start:doc-sc-metadata-relay]
    private suspend fun relayConversationPacket(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        val conversationId = dispatch.metadata()[SampleNames.ConversationIdMetadataKey]
        val target = conversationId?.let { conversationActors[it] } ?: requireIdentityActor()
        target.relay(dispatch, payload).await()
    }

    // --8<-- [end:doc-sc-metadata-relay]

    private fun requireIdentityActor(): ZLinkSessionActor =
        identityActor
            ?: throw IllegalStateException(
                "Client must authenticate before sending conversation packets."
            )

    private fun requireConversationId(dispatch: ZLinkSessionDispatchContext): String =
        dispatch.metadata()[SampleNames.ConversationIdMetadataKey]
            ?: throw IllegalStateException(
                "Conversation packet is missing the ConversationId metadata."
            )

    private companion object {
        private val logger = LoggerFactory.getLogger(SupportChatSession::class.java)
    }
}
